// @tsparticles/engine/browser/Core/RenderManager.js
import { defaultCompositeValue, defaultTransformValue, defaultZoom, minStrokeWidth, minimumSize, originPoint, zIndexFactorOffset, } from "./Utils/Constants.js";
import { getStyleFromHsl, rangeColorToHsl } from "../Utils/ColorUtils.js";
const fColorIndex = 0, sColorIndex = 1;
function setTransformValue(factor, newFactor, key) {
    const newValue = newFactor[key];
    if (newValue !== undefined) {
        factor[key] = (factor[key] ?? defaultTransformValue) * newValue;
    }
}
export class RenderManager {
    #canvasClearPlugins;
    #canvasManager;
    #canvasPaintPlugins;
    #clearDrawPlugins;
    #colorPlugins;
    #container;
    #context;
    #contextSettings;
    #drawParticlePlugins;
    #drawParticlesCleanupPlugins;
    #drawParticlesSetupPlugins;
    #drawPlugins;
    #drawSettingsCleanupPlugins;
    #drawSettingsSetupPlugins;
    #pluginManager;
    #postDrawUpdaters;
    #preDrawUpdaters;
    #reusableColorStyles = {};
    #reusablePluginColors = [undefined, undefined];
    #reusableTransform = {};
    constructor(pluginManager, container, canvasManager) {
        this.#pluginManager = pluginManager;
        this.#container = container;
        this.#canvasManager = canvasManager;
        this.#context = null;
        this.#preDrawUpdaters = [];
        this.#postDrawUpdaters = [];
        this.#colorPlugins = [];
        this.#canvasClearPlugins = [];
        this.#canvasPaintPlugins = [];
        this.#clearDrawPlugins = [];
        this.#drawParticlePlugins = [];
        this.#drawParticlesCleanupPlugins = [];
        this.#drawParticlesSetupPlugins = [];
        this.#drawPlugins = [];
        this.#drawSettingsSetupPlugins = [];
        this.#drawSettingsCleanupPlugins = [];
    }
    get settings() {
        return this.#contextSettings;
    }
    canvasClear() {
        if (!this.#container.actualOptions.clear) {
            return;
        }
        this.draw(ctx => {
            ctx.clearRect(originPoint.x, originPoint.y, this.#canvasManager.size.width, this.#canvasManager.size.height);
        });
    }
    clear() {
        let pluginHandled = false;
        for (const plugin of this.#canvasClearPlugins) {
            pluginHandled = plugin.canvasClear?.() ?? false;
            if (pluginHandled) {
                break;
            }
        }
        if (pluginHandled) {
            return;
        }
        this.canvasClear();
    }
    destroy() {
        this.stop();
        this.#preDrawUpdaters = [];
        this.#postDrawUpdaters = [];
        this.#colorPlugins = [];
        this.#canvasClearPlugins = [];
        this.#canvasPaintPlugins = [];
        this.#clearDrawPlugins = [];
        this.#drawParticlePlugins = [];
        this.#drawParticlesCleanupPlugins = [];
        this.#drawParticlesSetupPlugins = [];
        this.#drawPlugins = [];
        this.#drawSettingsSetupPlugins = [];
        this.#drawSettingsCleanupPlugins = [];
    }
    draw(cb) {
        const ctx = this.#context;
        if (!ctx) {
            return;
        }
        return cb(ctx);
    }
    drawParticle(particle, delta) {
        if (particle.spawning || particle.destroyed) {
            return;
        }
        const radius = particle.getRadius();
        if (radius <= minimumSize) {
            return;
        }
        const pfColor = particle.getFillColor(), psColor = particle.getStrokeColor();
        let [fColor, sColor] = this.#getPluginParticleColors(particle);
        fColor ??= pfColor;
        sColor ??= psColor;
        if (!fColor && !sColor) {
            return;
        }
        const container = this.#container, zIndexOptions = particle.options.zIndex, zIndexFactor = zIndexFactorOffset - particle.zIndexFactor, { fillOpacity, opacity, strokeOpacity } = particle.getOpacity(), transform = this.#reusableTransform, colorStyles = this.#reusableColorStyles, fill = fColor ? getStyleFromHsl(fColor, container.hdr, fillOpacity * opacity) : undefined, stroke = sColor ? getStyleFromHsl(sColor, container.hdr, strokeOpacity * opacity) : fill;
        transform.a = transform.b = transform.c = transform.d = undefined;
        colorStyles.fill = fill;
        colorStyles.stroke = stroke;
        this.draw((context) => {
            for (const plugin of this.#drawParticlesSetupPlugins) {
                plugin.drawParticleSetup?.(context, particle, delta);
            }
            this.#applyPreDrawUpdaters(context, particle, radius, opacity, colorStyles, transform);
            this.#drawParticle({
                container,
                context,
                particle,
                delta,
                colorStyles,
                radius: radius * zIndexFactor ** zIndexOptions.sizeRate,
                opacity: opacity,
                transform,
            });
            this.#applyPostDrawUpdaters(particle);
            for (const plugin of this.#drawParticlesCleanupPlugins) {
                plugin.drawParticleCleanup?.(context, particle, delta);
            }
        });
    }
    drawParticlePlugins(particle, delta) {
        this.draw(ctx => {
            for (const plugin of this.#drawParticlePlugins) {
                this.#drawParticlePlugin(ctx, plugin, particle, delta);
            }
        });
    }
    drawParticles(delta) {
        const { particles } = this.#container;
        this.clear();
        particles.update(delta);
        this.draw(ctx => {
            for (const plugin of this.#drawSettingsSetupPlugins) {
                plugin.drawSettingsSetup?.(ctx, delta);
            }
            for (const plugin of this.#drawPlugins) {
                plugin.draw?.(ctx, delta);
            }
            particles.drawParticles(delta);
            for (const plugin of this.#clearDrawPlugins) {
                plugin.clearDraw?.(ctx, delta);
            }
            for (const plugin of this.#drawSettingsCleanupPlugins) {
                plugin.drawSettingsCleanup?.(ctx, delta);
            }
        });
    }
    init() {
        this.initUpdaters();
        this.initPlugins();
        this.paint();
    }
    initPlugins() {
        this.#colorPlugins = [];
        this.#canvasClearPlugins = [];
        this.#canvasPaintPlugins = [];
        this.#clearDrawPlugins = [];
        this.#drawParticlePlugins = [];
        this.#drawParticlesSetupPlugins = [];
        this.#drawParticlesCleanupPlugins = [];
        this.#drawPlugins = [];
        this.#drawSettingsSetupPlugins = [];
        this.#drawSettingsCleanupPlugins = [];
        for (const plugin of this.#container.plugins) {
            if (plugin.particleFillColor ?? plugin.particleStrokeColor) {
                this.#colorPlugins.push(plugin);
            }
            if (plugin.canvasClear) {
                this.#canvasClearPlugins.push(plugin);
            }
            if (plugin.canvasPaint) {
                this.#canvasPaintPlugins.push(plugin);
            }
            if (plugin.drawParticle) {
                this.#drawParticlePlugins.push(plugin);
            }
            if (plugin.drawParticleSetup) {
                this.#drawParticlesSetupPlugins.push(plugin);
            }
            if (plugin.drawParticleCleanup) {
                this.#drawParticlesCleanupPlugins.push(plugin);
            }
            if (plugin.draw) {
                this.#drawPlugins.push(plugin);
            }
            if (plugin.drawSettingsSetup) {
                this.#drawSettingsSetupPlugins.push(plugin);
            }
            if (plugin.drawSettingsCleanup) {
                this.#drawSettingsCleanupPlugins.push(plugin);
            }
            if (plugin.clearDraw) {
                this.#clearDrawPlugins.push(plugin);
            }
        }
    }
    initUpdaters() {
        this.#preDrawUpdaters = [];
        this.#postDrawUpdaters = [];
        for (const updater of this.#container.particleUpdaters) {
            if (updater.afterDraw) {
                this.#postDrawUpdaters.push(updater);
            }
            if (updater.getColorStyles ?? updater.getTransformValues ?? updater.beforeDraw) {
                this.#preDrawUpdaters.push(updater);
            }
        }
    }
    paint() {
        let handled = false;
        for (const plugin of this.#canvasPaintPlugins) {
            handled = plugin.canvasPaint?.() ?? false;
            if (handled) {
                break;
            }
        }
        if (handled) {
            return;
        }
        this.paintBase();
    }
    paintBase(baseColor) {
        this.draw(ctx => {
            ctx.fillStyle = baseColor ?? "rgba(0,0,0,0)";
            ctx.fillRect(originPoint.x, originPoint.y, this.#canvasManager.size.width, this.#canvasManager.size.height);
        });
    }
    paintImage(image, opacity) {
        this.draw(ctx => {
            if (!image) {
                return;
            }
            const prevAlpha = ctx.globalAlpha;
            ctx.globalAlpha = opacity;
            ctx.drawImage(image, originPoint.x, originPoint.y, this.#canvasManager.size.width, this.#canvasManager.size.height);
            ctx.globalAlpha = prevAlpha;
        });
    }
    setContext(context) {
        this.#context = context;
        if (this.#context) {
            this.#context.globalCompositeOperation = defaultCompositeValue;
        }
    }
    setContextSettings(settings) {
        this.#contextSettings = settings;
    }
    stop() {
        this.draw(ctx => {
            ctx.clearRect(originPoint.x, originPoint.y, this.#canvasManager.size.width, this.#canvasManager.size.height);
        });
    }
    #applyPostDrawUpdaters(particle) {
        for (const updater of this.#postDrawUpdaters) {
            updater.afterDraw?.(particle);
        }
    }
    #applyPreDrawUpdaters(ctx, particle, radius, zOpacity, colorStyles, transform) {
        for (const updater of this.#preDrawUpdaters) {
            if (updater.getColorStyles) {
                const { fill, stroke } = updater.getColorStyles(particle, ctx, radius, zOpacity);
                if (fill) {
                    colorStyles.fill = fill;
                }
                if (stroke) {
                    colorStyles.stroke = stroke;
                }
            }
            if (updater.getTransformValues) {
                const updaterTransform = updater.getTransformValues(particle);
                for (const key in updaterTransform) {
                    setTransformValue(transform, updaterTransform, key);
                }
            }
            updater.beforeDraw?.(particle);
        }
    }
    #drawAfterEffect(drawer, data) {
        if (!drawer?.drawAfter) {
            return;
        }
        const { particle } = data;
        if (!particle.effect) {
            return;
        }
        drawer.drawAfter(data);
    }
    #drawBeforeEffect(drawer, data) {
        if (!drawer?.drawBefore) {
            return;
        }
        const { particle } = data;
        if (!particle.effect) {
            return;
        }
        drawer.drawBefore(data);
    }
    #drawParticle(data) {
        const { container, context, particle, delta, colorStyles, radius, opacity, transform } = data, { effectDrawers, shapeDrawers } = container, pos = particle.getPosition(), transformData = particle.getTransformData(transform), drawScale = defaultZoom, drawPosition = {
            x: pos.x,
            y: pos.y,
        };
        context.setTransform(transformData.a, transformData.b, transformData.c, transformData.d, pos.x, pos.y);
        if (colorStyles.fill) {
            context.fillStyle = colorStyles.fill;
        }
        const fillEnabled = !!particle.fillEnabled, strokeWidth = particle.strokeWidth ?? minStrokeWidth;
        context.lineWidth = strokeWidth;
        if (colorStyles.stroke) {
            context.strokeStyle = colorStyles.stroke;
        }
        const drawData = {
            context,
            particle,
            radius,
            drawRadius: radius * drawScale,
            opacity,
            delta,
            pixelRatio: container.retina.pixelRatio,
            fill: fillEnabled,
            stroke: strokeWidth > minStrokeWidth,
            transformData,
            position: { ...pos },
            drawPosition,
            drawScale,
        };
        for (const plugin of container.plugins) {
            plugin.drawParticleTransform?.(drawData);
        }
        const effect = particle.effect ? effectDrawers.get(particle.effect) : undefined, shape = particle.shape ? shapeDrawers.get(particle.shape) : undefined;
        this.#drawBeforeEffect(effect, drawData);
        this.#drawShapeBeforeDraw(shape, drawData);
        this.#drawShape(shape, drawData);
        this.#drawShapeAfterDraw(shape, drawData);
        this.#drawAfterEffect(effect, drawData);
        context.resetTransform();
    }
    #drawParticlePlugin(context, plugin, particle, delta) {
        if (!plugin.drawParticle) {
            return;
        }
        plugin.drawParticle(context, particle, delta);
    }
    #drawShape(drawer, data) {
        if (!drawer) {
            return;
        }
        const { context, fill, particle, stroke } = data;
        if (!particle.shape) {
            return;
        }
        context.beginPath();
        drawer.draw(data);
        if (particle.shapeClose) {
            context.closePath();
        }
        if (fill) {
            context.fill();
        }
        if (stroke) {
            context.stroke();
        }
    }
    #drawShapeAfterDraw(drawer, data) {
        if (!drawer?.afterDraw) {
            return;
        }
        const { particle } = data;
        if (!particle.shape) {
            return;
        }
        drawer.afterDraw(data);
    }
    #drawShapeBeforeDraw(drawer, data) {
        if (!drawer?.beforeDraw) {
            return;
        }
        const { particle } = data;
        if (!particle.shape) {
            return;
        }
        drawer.beforeDraw(data);
    }
    #getPluginParticleColors(particle) {
        let fColor, sColor;
        for (const plugin of this.#colorPlugins) {
            if (!fColor && plugin.particleFillColor) {
                fColor = rangeColorToHsl(this.#pluginManager, plugin.particleFillColor(particle));
            }
            if (!sColor && plugin.particleStrokeColor) {
                sColor = rangeColorToHsl(this.#pluginManager, plugin.particleStrokeColor(particle));
            }
            if (fColor && sColor) {
                break;
            }
        }
        this.#reusablePluginColors[fColorIndex] = fColor;
        this.#reusablePluginColors[sColorIndex] = sColor;
        return this.#reusablePluginColors;
    }
}
